action/candidate_endorsement.go

```diff
@@ -146,28 +249,69 @@ func NewCandidateEndorsement(nonce, gasLimit uint64, gasPrice *big.Int, bucketIn
 	}
 }
 
+// NewCandidateEndorsement returns a CandidateEndorsement action
+func NewCandidateEndorsement(nonce, gasLimit uint64, gasPrice *big.Int, bucketIndex uint64, op CandidateEndorsementOp) (*CandidateEndorsement, error) {
+	if op == CandidateEndorsementOpLegacy {
+		return nil, errors.New("invalid operation")
+	}
+	return &CandidateEndorsement{
+		AbstractAction: AbstractAction{
+			version:  version.ProtocolVersion,
+			nonce:    nonce,
+			gasLimit: gasLimit,
+			gasPrice: gasPrice,
+		},
+		bucketIndex: bucketIndex,
+		op:          op,
+	}, nil
+}
+
 // NewCandidateEndorsementFromABIBinary parses the smart contract input and creates an action
 func NewCandidateEndorsementFromABIBinary(data []byte) (*CandidateEndorsement, error) {
+	if len(data) <= 4 {
+		return nil, errDecodeFailure
+	}
 	var (
 		paramsMap = map[string]any{}
 		cr        CandidateEndorsement
+		method    abi.Method
 	)
-	// sanity check
-	if len(data) <= 4 || !bytes.Equal(candidateEndorsementMethod.ID, data[:4]) {
+	switch {
+	case bytes.Equal(candidateEndorsementLegacyMethod.ID, data[:4]):
+		if err := candidateEndorsementLegacyMethod.Inputs.UnpackIntoMap(paramsMap, data[4:]); err != nil {
+			return nil, err
+		}
+		bucketID, ok := paramsMap["bucketIndex"].(uint64)
+		if !ok {
+			return nil, errDecodeFailure
+		}
+		endorse, ok := paramsMap["endorse"].(bool)
+		if !ok {
+			return nil, errDecodeFailure
+		}
+		cr.bucketIndex = bucketID
+		cr.endorse = endorse
+		cr.op = CandidateEndorsementOpLegacy
+		return &cr, nil
```

Comment:
see comments in 4312, better to also set `method` and `cr.op` and use common unpack code after `switch`